[apex] ANTLR runtime mismatch 4.9.1 used for code generation does not match the current runtime version 4.13.2

<!-- Please, prefix the report title with the language it applies to within brackets, such as [java] or [apex].
If not specific to a language, you can use [core]. -->

**Affects PMD Version:** 

Make sure, to test with the latest PMD version.

**Description:**

Cannot use multifile analysis on apex

**Exception Stacktrace:**

```
ANTLR Tool version 4.9.1 used for code generation does not match the current runtime version 4.13.2
ANTLR Runtime version 4.9.1 used for parser compilation does not match the current runtime version 4.13.2
[ERROR] Exception while initializing Apexlink (null)
java.lang.ExceptionInInitializerError
	at com.nawforce.runtime.parsers.PageParser.parse(PageParser.scala:52)
	at com.nawforce.runtime.parsers.PageParser.parsePage(PageParser.scala:35)
	at com.nawforce.pkgforce.stream.PageGenerator$.$anonfun$toEvents$1(PageGenerator.scala:49)
	at scala.Option.map(Option.scala:242)
	at com.nawforce.pkgforce.stream.PageGenerator$.toEvents(PageGenerator.scala:47)
	at com.nawforce.pkgforce.stream.PageGenerator$.$anonfun$iterator$2(PageGenerator.scala:40)
	at scala.Option.map(Option.scala:242)
	at com.nawforce.pkgforce.stream.PageGenerator$.$anonfun$iterator$1(PageGenerator.scala:40)
	at scala.collection.immutable.List.flatMap(List.scala:283)
	at scala.collection.immutable.List.flatMap(List.scala:79)
	at com.nawforce.pkgforce.stream.PageGenerator$.iterator(PageGenerator.scala:40)
	at com.nawforce.pkgforce.stream.PackageStream$.$anonfun$eventStream$2(PackageStream.scala:60)
	at scala.collection.Iterator$ConcatIteratorCell.headIterator(Iterator.scala:1220)
	at scala.collection.Iterator$ConcatIterator.advance$1(Iterator.scala:1184)
	at scala.collection.Iterator$ConcatIterator.hasNext(Iterator.scala:1195)
	at scala.collection.Iterator$$anon$1.hasNext(Iterator.scala:139)
	at com.nawforce.apexlink.org.package$.bufferEvents(package.scala:41)
	at com.nawforce.apexlink.org.StreamDeployer.consumeLabels(StreamDeployer.scala:81)
	at com.nawforce.apexlink.org.StreamDeployer.$anonfun$load$1(StreamDeployer.scala:57)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
	at com.nawforce.apexlink.types.platform.PlatformTypes$.withLoadingObserver(PlatformTypes.scala:67)
	at com.nawforce.apexlink.org.StreamDeployer.load(StreamDeployer.scala:55)
	at com.nawforce.apexlink.org.StreamDeployer.<init>(StreamDeployer.scala:48)
	at com.nawforce.apexlink.org.OPM$Module.freeze(OPM.scala:696)
	at com.nawforce.apexlink.org.OPM$OrgImpl.$anonfun$new$3(OPM.scala:164)
	at com.nawforce.apexlink.org.OPM$OrgImpl.$anonfun$new$3$adapted(OPM.scala:164)
	at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:630)
	at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:628)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:936)
	at com.nawforce.apexlink.org.OPM$OrgImpl.$anonfun$new$2(OPM.scala:164)
	at com.nawforce.apexlink.org.OPM$OrgImpl.$anonfun$new$2$adapted(OPM.scala:164)
	at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:630)
	at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:628)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:936)
	at com.nawforce.apexlink.org.OPM$OrgImpl.$anonfun$new$1(OPM.scala:164)
	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
	at com.nawforce.apexlink.org.OPM$OrgImpl.<init>(OPM.scala:163)
	at com.nawforce.apexlink.api.Org$.$anonfun$newOrg$9(Org.scala:295)
	at com.nawforce.pkgforce.diagnostics.LoggerOps$.time(Logger.scala:125)
	at com.nawforce.pkgforce.diagnostics.LoggerOps$.infoTime(Logger.scala:106)
	at com.nawforce.apexlink.api.Org$.newOrg(Org.scala:293)
	at com.nawforce.apexlink.api.Org$.newOrg(Org.scala:257)
	at com.nawforce.apexlink.api.Org$.newOrg(Org.scala:250)
	at com.nawforce.apexlink.api.Org.newOrg(Org.scala)
	at net.sourceforge.pmd.lang.apex.multifile.ApexMultifileAnalysis.<init>(ApexMultifileAnalysis.java:78)
	at net.sourceforge.pmd.lang.apex.multifile.InternalApiBridge.createApexMultiFileAnalysis(InternalApiBridge.java:26)
	at net.sourceforge.pmd.lang.apex.ApexLanguageProcessor.<init>(ApexLanguageProcessor.java:22)
	at net.sourceforge.pmd.lang.apex.ApexLanguageModule.createProcessor(ApexLanguageModule.java:47)
	at net.sourceforge.pmd.lang.LanguageProcessorRegistry.create(LanguageProcessorRegistry.java:143)
	at net.sourceforge.pmd.lang.LanguageProcessorRegistry.create(LanguageProcessorRegistry.java:120)
	at net.sourceforge.pmd.PmdAnalysis.performAnalysisImpl(PmdAnalysis.java:427)
	at net.sourceforge.pmd.PmdAnalysis.performAnalysisImpl(PmdAnalysis.java:382)
	at net.sourceforge.pmd.PmdAnalysis.performAnalysis(PmdAnalysis.java:362)
	at net.sourceforge.pmd.PmdAnalysis.runAndReturnStats(PmdAnalysis.java:614)
	at net.sourceforge.pmd.cli.commands.internal.PmdCommand.doExecute(PmdCommand.java:318)
	at net.sourceforge.pmd.cli.commands.internal.PmdCommand.doExecute(PmdCommand.java:51)
	at net.sourceforge.pmd.cli.internal.PmdRootLogger.executeInLoggingContext(PmdRootLogger.java:55)
	at net.sourceforge.pmd.cli.commands.internal.AbstractAnalysisPmdSubcommand.execute(AbstractAnalysisPmdSubcommand.java:188)
	at net.sourceforge.pmd.cli.commands.internal.AbstractPmdSubcommand.call(AbstractPmdSubcommand.java:30)
	at net.sourceforge.pmd.cli.commands.internal.AbstractPmdSubcommand.call(AbstractPmdSubcommand.java:16)
	at picocli.CommandLine.executeUserObject(CommandLine.java:2031)
	at picocli.CommandLine.access$1500(CommandLine.java:148)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2469)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2461)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2423)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2277)
	at picocli.CommandLine$RunLast.execute(CommandLine.java:2425)
	at picocli.CommandLine.execute(CommandLine.java:2174)
	at net.sourceforge.pmd.cli.PmdCli.mainWithoutExit(PmdCli.java:25)
	at net.sourceforge.pmd.cli.PmdCli.main(PmdCli.java:29)
Caused by: java.lang.UnsupportedOperationException: java.io.InvalidClassException: org.antlr.v4.runtime.atn.ATN; Could not deserialize ATN with version 3 (expected 4).
	at org.antlr.v4.runtime.atn.ATNDeserializer.deserialize(ATNDeserializer.java:56)
	at org.antlr.v4.runtime.atn.ATNDeserializer.deserialize(ATNDeserializer.java:48)
	at com.nawforce.vfparser.VFLexer.<clinit>(VFLexer.java:303)
	... 70 more
Caused by: java.io.InvalidClassException: org.antlr.v4.runtime.atn.ATN; Could not deserialize ATN with version 3 (expected 4).
	... 73 more
[ERROR] PMD will not attempt to initialize Apexlink further, this can cause rules like UnusedMethod and AvoidInterfaceAsMapKey to be dysfunctional
```

**Code Sample demonstrating the issue:**

```java

```

**Steps to reproduce:**

Please provide detailed steps for how we can reproduce the bug.

1.  pmd check --rulesets category/apex/design.xml --dir ./src/folder/classes/Foo.cls


**Running PMD through:** *[CLI]*

<!-- If relevant, also include your JDK and OS information, e.g. for ClassNotFoundException, LinkageError, reflection failures, etc. -->
openjdk 25.0.3 2026-04-21 LTS

PMD 7.25.0 (418f8b7413a1c0ecb3ee409b6edeedeae5c8df39, 2026-05-29T06:28:38Z)
Java version: 25.0.3, vendor: Amazon.com Inc., runtime: C:\Program Files\Amazon Corretto\jdk25.0.3_9

also occurs on PMD 7.26.0-SNAPSHOT (a91430bc98b3d7e5d68416a2f9caf5e5634940a9, 2026-06-22T11:29:46Z)

--- a/docs/pages/release_notes.md
+++ b/docs/pages/release_notes.md
@@ -38,6 +38,8 @@ This is a {{ site.pmd.release_type }} release.
   and may lead to confusion about which value is used.
 
 ### 🐛️ Fixed Issues
+* apex
+  * [#6806](https://github.com/pmd/pmd/issues/6806): \[apex] ANTLR runtime mismatch 4.9.1 used for code generation does not match the current runtime version 4.13.2
 * apex-security
   * [#2955](https://github.com/pmd/pmd/issues/2955): \[apex] ApexSOQLInjection: False positive when passing local var with concatenating strings
   * [#3877](https://github.com/pmd/pmd/issues/3877): \[apex] ApexCRUDViolation: False positive with Lists of Objects with getSObjectType().getDescribe()

--- a/pmd-apex/pom.xml
+++ b/pmd-apex/pom.xml
@@ -108,6 +108,11 @@
             <artifactId>apex-ls_2.13</artifactId>
             <version>6.0.2</version>
         </dependency>
+        <dependency>
+            <groupId>io.github.apex-dev-tools</groupId>
+            <artifactId>vf-parser</artifactId>
+            <version>2.0.0-beta.1</version>
+        </dependency>
 
         <dependency>
             <groupId>org.apache.commons</groupId>

--- a/pmd-apex/src/test/resources/net/sourceforge/pmd/lang/apex/rule/design/UnusedMethod/project1/src/SomePage.page
+++ b/pmd-apex/src/test/resources/net/sourceforge/pmd/lang/apex/rule/design/UnusedMethod/project1/src/SomePage.page
@@ -1,0 +1,5 @@
+<apex:page>
+The contents of this page aren't relevant to the test, but the test requires the file to be present.
+This causes apex-link to parse this file via vf-parser. This detects issues like https://github.com/pmd/pmd/issues/6806
+when the ANTLR runtime doesn't match the compile-time version of vf-parser.
+</apex:page>

